### PR TITLE
Publish server helm chart as part of CI

### PR DESCRIPTION
**Pull request by wjayesh**
## Describe changes
I implemented a gha to publish the zenml server helm chart to an ECR repo.

## Pre-requisites
Please ensure you have done the following:
- [ ] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: .github/workflows/publish_helm_chart.yaml [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# package and push the zenml helm chart to the ECR repository
+name: Publish Helm Chart
+
+on: workflow_call
```
Comment (AlexejPenner) [inline .github/workflows/publish_helm_chart.yaml:4]: Might make sense to also allow this to be called from workflow dispatch in case it fails on release

File: .github/workflows/publish_helm_chart.yaml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      PYTHONIOENCODING: 'utf-8'
+
+    steps:
+      - name: Checkout repo
+        uses: actions/checkout@v3
+
+      # add tag version to Chart.yaml file for variable version
+      - name: Add tag version to Chart.yaml
+        run: |
+          sed -i "s/version: .*/version: \"${{ github.ref_name }}\"/" src/zenml/zen_server/deploy/helm/Chart.yaml
```
Comment (AlexejPenner) [inline .github/workflows/publish_helm_chart.yaml:22]: When it comes to commands like this, I like it when they are commented a bit verbosly:

# Finds like that contains "version: " and replaces the complete line with "version: ___"

File: src/zenml/zen_server/deploy/helm/Chart.yaml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 apiVersion: v2
-name: zenml
-version: "1.0"
+name: zenml-server
+version: "0.42.1"
```
Comment (AlexejPenner) [inline src/zenml/zen_server/deploy/helm/Chart.yaml:3]: SHouldn't this be 0.43.0 already? How do we do it moving forward. DOes this get manually updated as well? If Not, could it make sense to leave the version number empty?

**Review by AlexejPenner — COMMENTED** [review]
Got a few suggestions/ thoughts

File: .github/workflows/publish_helm_chart.yaml (hunk 2 above)
Comment (AlexejPenner) [reply .github/workflows/publish_helm_chart.yaml:22]: Didin't mean to scream it :D

File: .github/workflows/publish_helm_chart.yaml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+      - name: Configure AWS credentials
+        uses: aws-actions/configure-aws-credentials@v2
+        with:
+          role-to-assume: arn:aws:iam::715803424590:role/gh-action-role-zenml-helm
```
[the remark this answers is not in the document]
Comment (wjayesh) [reply .github/workflows/publish_helm_chart.yaml:27]: yes, exactly

File: src/zenml/zen_server/deploy/helm/Chart.yaml (hunk 3 above)
Comment (wjayesh) [reply src/zenml/zen_server/deploy/helm/Chart.yaml:3]: yes, it does get updated autonomously as part of the CI. A step fetches the latest verison and updates this file before it is packaged. You're right, it makes sense to leave it empty; I'll make the change.

File: .github/workflows/publish_helm_chart.yaml (hunk 2 above)
Comment (wjayesh) [reply .github/workflows/publish_helm_chart.yaml:29]: ahaha, point taken

File: src/zenml/zen_server/deploy/helm/templates/hpa.yaml [hunk 5]
```diff
@@ -1,5 +1,5 @@
 {{- if .Values.autoscaling.enabled }}
-apiVersion: autoscaling/v2beta1
+apiVersion: autoscaling/v2
```
Comment (stefannica) [inline src/zenml/zen_server/deploy/helm/templates/hpa.yaml:2]: What implications does this have on the supported Kuberentes cluster versions ? Should this be done with an if/else depending on that kube API version ?

File: src/zenml/zen_server/deploy/helm/templates/hpa.yaml [hunk 6]
```diff
@@ -17,12 +17,16 @@ spec:
     - type: Resource
       resource:
         name: cpu
-        targetAverageUtilization: {{ .Values.autoscaling.targetCPUUtilizationPercentage }}
+        target:
+          type:  Utilization
+          averageUtilization: {{ .Values.autoscaling.targetCPUUtilizationPercentage }}
```
Comment (stefannica) [inline src/zenml/zen_server/deploy/helm/templates/hpa.yaml:22]: Same question here: is this change required because some Kubernetes cluster APIs don't support the previous syntax ? If so, shouldn't it be in an if/else condition to maintain backwards compatibility with previous versions ?

**Review by stefannica — CHANGES_REQUESTED** [review]

File: src/zenml/zen_server/deploy/helm/Chart.yaml [hunk 7]
```diff
@@ -1,6 +1,6 @@
 apiVersion: v2
-name: zenml
-version: "1.0"
+name: zenml-server
```
[the remark this answers is not in the document]
Comment (wjayesh) [reply src/zenml/zen_server/deploy/helm/Chart.yaml:2]: No strong reason. I felt that this reflected what the helm chart deploys, more appropriately, since it is now in the public light more than ever. Let me know if I should consider going back

File: src/zenml/zen_server/deploy/helm/templates/hpa.yaml (hunk 5 above)
Comment (wjayesh) [reply src/zenml/zen_server/deploy/helm/templates/hpa.yaml:2]: With this change, we would stop supporting clusters with versions starting 1.23 and older. I think it's reasonable to no longer support that since 1.23 has reached end of support almost 6 months back and most cloud providers use 1.27 as the default cluster version.

File: src/zenml/zen_server/deploy/helm/templates/hpa.yaml (hunk 6 above)
Comment (wjayesh) [reply src/zenml/zen_server/deploy/helm/templates/hpa.yaml:22]: Yes, the new syntax is need for v2 of the HPA resource. The older versions would no longer be supported. What I can do is add a different chart version that has support for v1beta1 for the same zenml version but that might need some changes across the board. Still the safest bet and allows us to support everyone.

File: src/zenml/zen_server/deploy/helm/Chart.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (stefannica) [reply src/zenml/zen_server/deploy/helm/Chart.yaml:2]: No need to go back, just considering the implications in the name change. No idea what this does for current deployments.

File: src/zenml/zen_server/deploy/helm/templates/hpa.yaml (hunk 6 above)
Comment (stefannica) [reply src/zenml/zen_server/deploy/helm/templates/hpa.yaml:22]: You don't need a different chart version, just an if condition here. See this thread for example: https://stackoverflow.com/questions/70346448/deploy-conditional-charts-based-on-kubeversion-helm

**Review by stefannica — APPROVED** [review]
Fine to merge it as is as long as we're not breaking existing deployments.
